gateway/broadcaster/broadcaster.go

```diff
@@ -0,0 +1,123 @@
+package broadcaster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	pb "github.com/spacemeshos/api/release/go/spacemesh/v1"
+	"github.com/spacemeshos/smutil/log"
+	"google.golang.org/genproto/googleapis/rpc/code"
+	"google.golang.org/grpc"
+)
+
+const DefaultBroadcastTimeout = 30 * time.Second
+
+// Broadcaster is responsible for broadcasting proofs to a list of Spacemesh-compatible gateway nodes.
+type Broadcaster struct {
+	clients               []pb.GatewayServiceClient
+	connections           []*grpc.ClientConn
+	broadcastTimeout      time.Duration
+	broadcastAckThreshold uint
+}
+
+// New instantiate a new Broadcaster for a given list of gateway nodes addresses.
+// disableBroadcast allows to create a disabled Broadcaster instance.
+// connTimeout set the timeout per gRPC connection attempt to a node.
+// broadcastTimeout set the timeout per proof broadcast.
+// broadcastAcksThreshold set the lower-bound of required successful proof broadcasts. If not met, a warning will be logged.
+func New(connections []*grpc.ClientConn, disableBroadcast bool, broadcastTimeout time.Duration, broadcastAcksThreshold uint) (*Broadcaster, error) {
+	if disableBroadcast {
+		log.Info("Broadcast is disabled")
+		return &Broadcaster{}, nil
+	}
+
+	if broadcastAcksThreshold < 1 {
+		return nil, errors.New("successful broadcast threshold must be greater than 0")
+	}
+	if len(connections) < int(broadcastAcksThreshold) {
+		return nil, fmt.Errorf("number of gateway connections (%d) must be greater or equal than the successful broadcast threshold (%d)", len(connections), broadcastAcksThreshold)
+	}
+
+	clients := make([]pb.GatewayServiceClient, len(connections))
+	for i, conn := range connections {
+		clients[i] = pb.NewGatewayServiceClient(conn)
+	}
+
+	return &Broadcaster{
+		clients:               clients,
+		connections:           connections,
+		broadcastTimeout:      broadcastTimeout,
+		broadcastAckThreshold: broadcastAcksThreshold,
+	}, nil
+}
+
+// BroadcastProof broadcasts a serialized proof of a given round.
+func (b *Broadcaster) BroadcastProof(msg []byte, roundID string, members [][]byte) error {
+	if b.clients == nil {
+		log.Info("Broadcast is disabled, not broadcasting round %v proof", roundID)
+		return nil
+	}
+	pbMsg := &pb.BroadcastPoetRequest{Data: msg}
+	ctx, cancel := context.WithTimeout(context.Background(), b.broadcastTimeout)
+	defer cancel()
+
+	responses := make([]*pb.BroadcastPoetResponse, len(b.clients))
+	errs := make([]error, len(b.clients))
+
+	start := time.Now()
+	var wg sync.WaitGroup
+	wg.Add(len(b.clients))
```

Comment:
This code is similar to the one used for connecting in the gateway manager. I prefer this approach with `WaitGroup` over `errGroup` and channels; it seems more straight forward to me.

Would it be much effort to adapt the other code to also use a wait group? 🙂

Reply:
I generally prefer error groups as they are more robust (there's no need to remember to call `.Add()` and `.Done()`).

I don't agree that the broadcaster's approach is more straightforward. Instead of consumers with a clear purpose, there is the nasty loop that splits results into OKs and errors, which, IMO, is bug-prone.

Reply:
I'm not sure I can follow your argument.

`errGroup.Group` has the advantage to `sync.WaitGroup` of automatically including cancellation of ongoing processes if one of the routines fails.

If you explicitly do not want to do this, the `sync.WaitGroup` seems to me a much more straight forward approach.

Reply:
My point is that `errogroup` has the advantage that I don't need to manually call `.Add()` and `.Done()` to count the running goroutines.

Reply:
I do understand that, but instead it has it's own context that we don't use here and you have to return an error with every routine that is started in the `errgroup` although it will always be `nil`.

So you you are swapping `.Add()` and `.Done()` for a `return nil` and internal functionality that isn't even used.